juju/juju
# Wireup parallel actions

The parallel and execution group attributes are used when running unit and machine actions.

Machine actions are run in parallel by default in separate go routines. The machine agent lock is wired through and a named lock is acquired if needed for non-parallel actions. The lock name is unique to running actions so does not affect the other users of the machine lock.
In addition, the relevant unit tests were reworked to use gomock since this allows much easier testing of the parallel execution.
The worker which runs the actions no longer exits if there's an error beginning or running an action - the action is marked as failed, allowing other actions to still run.

Unit actions also use the parallel and execution group to determine if the machine lock should be acquired. Note that for a given unit agent, actions for that agent are still run sequentially, since the state machine worker picks of the next task (hook, action etc) one at a time. But if units are co-located on the same machine, actions for these different units are no longer restricted to running sequentially globally for the machine.
The uniter changes included having to load the action to run before the execution started, not just pass in the action id. This was so that the action could be inspected to see if the machine lock was needed when setting up the executor.

## QA steps

bootstrap and deploy a simple charm with a couple of actions, one marked as parallel, the other not.
The action script should be something like
`for i in {1..10}; do echo "x"; sleep 1; done`

Add a second unit of the charm to the same machine, eg juju add-unit foo --to 1

Execute parallel actions on the two units

juju run foo/0 --background parallel-action && juju run foo/1 --background parallel-action
juju operations
juju show-task X

The start times of both actions should be the same.
Do the same but with an action without parallel=true in the metadata.
The second action should show a start time after the first one finishes.

Test juju exec similarly - at the moment, all juju exec scripts now run in parallel.

juju exec --machine 0 "for i in {1..10}; do echo "x"; sleep 1; done"  \
&& juju exec --machine 0 "for i in {1..10}; do echo "x"; sleep 1; done"

juju operations --machines 0
juju show-task X
